### PR TITLE
Added Bind method to SchedulerExtender interface

### PR DESCRIPTION
**Pull request by ravigadde**
- Bind is used to let the extender know of the scheduling decision
- Extender uses this to set aside resources against the pod

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by ravigadde** [comment]
cc @k82cn @davidopp @timstclair @sarat-k

File: plugin/pkg/scheduler/generic_scheduler.go
Comment (k82cn) [inline plugin/pkg/scheduler/generic_scheduler.go]: I'm thinking to implement `Binder` interface; HTTPBinder & DefaultBinder (call apiserver directly). We can put implementation in `HTTPExtender.go`.

**Review by k82cn — COMMENTED** [review]
Great!! Just check the interface :). I'll check it more today.

File: plugin/pkg/scheduler/algorithm/scheduler_interface.go
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/scheduler_interface.go]: Prefer to following interface: 

```
Bind(pod *v1.Pod, nodeCandidates []*v1.Node) error
```

In extender, the `node` maybe invalid. If only one node, extender can not re-schedule to other node because unknown scheduler's predicates; it may introduce more community between scheduler & extender.

`nodeCandidates` are prepared by scheduler, so extender search the list to find a vaoid  node. But it seems make affinity/ant-affinity hard to consider assumed Pods, let me check the code and back to you today.

Comment (ravigadde) [inline plugin/pkg/scheduler/algorithm/scheduler_interface.go]: Filter does that job. By the time it gets to Bind, node selection is done. The sequence of calls is 1) Filter 2) Prioritize and 3) Bind

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/scheduler_interface.go]: Here's the case in my mind:

1. S->E: Filter request: n1, n2, n3, n4
1. E->S: n1, n2, n3
1. S->E: prioritize request: n1, n2, n3
1. E->S: n1, n2, n3
1. E: n1 is not valid, e.g. revoked by Mesos
1. S->E: bind Pod to n1
1. E->S: reject

If we changed step 6 to `S-E: bind Pod to candidates: n1, n2, n3`, extender will return successful at step 7, e.g. `E-S: bind to n2`. But the problem is that: the assumed pods are also considered for affinity/anti-affnity; if we assumed all candidates (e.g. n1, n2, n3), it'll make scheduler slower.

As a conclusion, I'm OK to keep current interface to send only one host: step 5 is less case compared to normal task launching :).

Comment (ravigadde) [inline plugin/pkg/scheduler/algorithm/scheduler_interface.go]: Yes, this behavior is no different with k8s apiserver. If the node became unschedulable after the scheduler selected it, the binding can fail and the pod is scheduled again.

File: plugin/pkg/scheduler/core/extender.go
Comment (wanghaoran1988) [inline plugin/pkg/scheduler/core/extender.go]: Why there is a nil, shouldn't the extender return something?

Comment (ravigadde) [inline plugin/pkg/scheduler/core/extender.go]: I am assuming a relevant http error code is adequate. Lmk if you think an explicit error is better.

Comment (wanghaoran1988) [inline plugin/pkg/scheduler/core/extender.go]: It's ok is you use the response.status and response.statuscode, but the sent func will always return nil if your result param is null, that means the bind always succeed if the http request succeed.

File: plugin/pkg/scheduler/core/extender.go
Comment (wanghaoran1988) [inline plugin/pkg/scheduler/core/extender.go]: I am saying here, if the result is nil, always succeed.

**Comment by davidopp** [comment]
@kubernetes/sig-scheduling-pr-reviews

File: plugin/pkg/scheduler/core/extender.go
Comment (ravigadde) [inline plugin/pkg/scheduler/core/extender.go]: Thanks @wanghaoran1988 I missed that. Added a check and an integration test case for bind error.

**Comment by timothysc** [comment]
@ravigadde Do you really want to get this in for 1.6?

**Comment by ravigadde** [comment]
@timothysc Yes, it simplifies the extender implementation quite a bit. I am waiting for #41119 to be merged to refresh this PR.

**Comment by k82cn** [comment]
@timothysc , that'll be great if in 1.6 for kube-mesos :).

File: test/integration/scheduler/extender_test.go
Comment (timothysc) [inline test/integration/scheduler/extender_test.go]: More of a nit, but you have repeating blocks that can be pulled out above the 1st if.

File: test/integration/scheduler/extender_test.go
Comment (timothysc) [inline test/integration/scheduler/extender_test.go]: Generally hard sleeps are discouraged in the tests if they can be avoided.

**Review by timothysc — CHANGES_REQUESTED** [review]

File: plugin/pkg/scheduler/api/types.go
Comment (ravigadde) [inline plugin/pkg/scheduler/api/types.go]: The convention today is that this combines both the extender's intent to handle this method and the path where it is handled. If an extender didn't care about handling Bind OR if its an existing extender that doesn't handle binds at all, this is not populated. It helps preserve backward compatibility also.

File: test/integration/scheduler/extender_test.go
Comment (ravigadde) [inline test/integration/scheduler/extender_test.go]: I will make another pass once the other PR is merged. I wanted to quickly put something together so folks get an idea on how much of a change this is (per the last sig discussion).

**Comment by davidopp** [comment]
ping -- there's approaching zero time left to get this into 1.6 -- need to act quickly

**Comment by ravigadde** [comment]
Rebased and updated based on feedback.

PTAL @davidopp @timstclair @sarat-k

File: plugin/pkg/scheduler/core/extender.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/extender.go]: Doesn't this still need to be fixed?  Adding the "result"

Comment (ravigadde) [inline plugin/pkg/scheduler/core/extender.go]: There is no data to be returned. Is http code adequate or should I add an explicit error? I assumed http error code is adequate.

File: plugin/pkg/scheduler/core/extender.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/extender.go]: It seems worth while to report with result success or failure.

Comment (ravigadde) [inline plugin/pkg/scheduler/core/extender.go]: ok added an explicit response. Will push after integration test passes.

**Comment by davidopp** [comment]
@ravigadde if you want this into 1.6 please file an exception (details were in the email sent to kubernetes-dev a few days ago)

**Comment by ravigadde** [comment]
Added an explicit error message.

@timothysc PTAL

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: the logic for multiple extenders seems odd.  
what if 1 succeeds and the next fails?

Comment (timothysc) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Is there a use case for > 1?

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: I'd suggest to implement `Binder` interface (let extender bind to apiserver) and replace current default binder (bind to apiserver) according to configuration. There's a case that: extender.Bind succeed but bind to apiserver failed. The extender did not how to recover it. If extender sync up periodically, there'll be a race condition with scheduler's assumed timeout.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: I haven't run in to >1 use case but here is how extenders recover. On any Bind failure, the pod is not scheduled and scheduler will attempt to schedule the pod again a little bit later. This results in another Bind on the same pod UID. Extenders that have reserved resources in the earlier incarnation can free them up when that happens.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Please see the above response. The goal is to make minimal change to existing scheduler-apiserver interface.

Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: But how about Bind succeed but bind to apiserver failed? How to recover it?

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: It is the same recovery path. k8s scheduler will schedule the pod again which will result in another Bind for the same pod UID. That can be used by the extender to free up resources allocated earlier (or reuse them if the pod is being bound the same node).

Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: If so, the extender has to cache pod Id or reconcile task from resource manager.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Could you please add a comment on that.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Yes, extender was meant to be a scheduler of resources outside of kubernetes. Tracking the usage of such resources is left to the extender implementation.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Done.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: extender is only an adaptor in some case, e.g. kube-mesos, such kind of cache make adaptor complex.

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: That is a bigger change IMO and doesn't work with multiple extenders. So far, the extender implementation has left the primary interaction between scheduler and apiserver untouched.

Is your goal to start kubelet on the node where the pod is scheduled? It can be achieved by watching the apiserver for pod.Spec.NodeName != ""?

**Comment by ravigadde** [comment]
@timothysc I asked for a 1 day extension for this. Do you believe that is sufficient time? Are there any open issues from your perspective?

**Comment by timothysc** [comment]
I'm ok with it provided the comment + @k82cn is ok.

**Review by timothysc — APPROVED** [review]

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: @k82cn I'd like your final say here before we green light this.. 

u2 are the primary users of this feature.

Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: @ravigadde , it not about kubelet start. In kube-mesos refactor, I try to avoid cache which makes it complex. And I prefer no interaction (e.g. recover/re-schedule) between scheduler & extender; predicates/prefer are standalone, no interaction with other part, e.g. apiserver bind failed.

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: @k82cn 
Can you please restate your requirements from the extender? I understand you want to make it stateless, but I am not clear what you are trying to do.
How do you handle bind failures now, before this change?

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: BTW, this is from the issue #41235:

Currently, the extender allows an external process to filter and prioritize nodes when scheduling a pod. But the process may also need to do preparation before binding Pod to host; for example, in kube-mesos, the k8sm-scheduler will ask Mesos to start kubelet before binding the Pod to host

Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: > How do you handle bind failures now, before this change?

Currently, it will mark task as FAILED if bind apiserver failed.
In kube-mesos, it'll let tasks to do apiserver bind, so: if failed to create task, not bind; if failed to bind, task will mark as FAILED. No cache in kube-mesos.

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Can we have a quick chat in the interest of closing this soon? Please find me on slack at ravi.gadde

Comment (ravigadde) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: > Currently, it will mark task as FAILED if bind apiserver failed.

How does kube-mesos detect that scheduler's bind with apiserver failed?

Is slack blocked in China? Please suggest a method to have a quick chat.

**Comment by ravigadde** [comment]
@davidopp @timothysc @k82cn 

Talked to Klaus offline and here is the summary:

In our use case (and what this PR implements), Bind is used by the extender to set aside resources for the pod. The extender can deal with subsequent k8s scheduler->API server binding failure.

His requirement is slightly different in that he wants Bind to be offloaded to the extender. The extender he is thinking about is stateless and that is the primary motivation for offloading the Bind. The extender can reserve resources and make the binding with API server atomically without maintaining state.

But implementing Bind that way breaks multiple extenders (to which extender would we offload the bind?). There are three choices:
1) Drop support for multiple extenders and implement Bind the way he wants it. Our extender can work that way too. Not sure if anyone uses multiple extenders.
2) Check that only one extender can implement Bind method and offload Bind to that extender.
3) Keep Bind the same way as implemented in this PR and invent another mechanism to satisfy his use case.

Let me know your opinion on how you would like to proceed.

**Comment by timothysc** [comment]
> 1) Drop support for multiple extenders and implement Bind the way he wants it. Our extender can work that way too. Not sure if anyone uses multiple extenders.

Doubtful, which was a question I asked earlier.

> 2) Check that only one extender can implement Bind method and offload Bind to that extender.

I'm not terribly keen on that, b/c bind is at a different layer higher up the stack in the scheduler.

> 3) Keep Bind the same way as implemented in this PR and invent another mechanism to satisfy his use case.

At this point in the game that may be the only recourse.  @k82cn thoughts?  We need to settle on this today! 

> @timothysc, In Openshift, would Openshift allow an extender to post new Binding for a pod? 

I just started last week at Heptio, but we had no plans of using the extender mechanism in openshift.

**Comment by k82cn** [comment]
The major concern is how to handle out of sync between scheduler & extender because of `bind`. I'm OK to merge this in 1.6 for @ravigadde 's case; but it need some improvement for kube-mesos. Let's continue the discussion after 1.6.

**Comment by timothysc** [comment]
Is there a use case for multiple extenders?

**Comment by k82cn** [comment]
> Is there a use case for multiple extenders?

No from me; if necessary, the extender can handle it: aggregate them into one resource. Not all service implemented extender interface, so an adaptor is always necessary.

**Comment by timothysc** [comment]
@ravigadde waiting on final comment re: Is there a use case for multiple extenders? 

If we can move push it down to 1, then lgtm today that would be great.

**Comment by ravigadde** [comment]
@timothysc I haven't see anyone use multiple extenders. Are you suggesting pursuing Option 1?

**Comment by timothysc** [comment]
> Are you suggesting pursuing Option 1? 

Yes.

**Comment by ravigadde** [comment]
OK I am on it

**Comment by ethernetdan** [comment]
A code freeze exception has been submitted for this PR - please let @kubernetes/release-team know of any status changes.

**Comment by timothysc** [comment]
emails have already been sent.

**Comment by ravigadde** [comment]
@timothysc @k82cn 

Updated the PR with Option 1. Rebased and squashed, as the implementation is very different now

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Not yours; but I'd like to skip it or log if `err != nil`. I'm OK to keep it as is in this PR.

File: plugin/pkg/scheduler/factory/factory.go
Comment (k82cn) [inline plugin/pkg/scheduler/factory/factory.go]: Honestly, I prefer not to use `nil` to check that :). But no solution, let's keep it fro this PR.

**Review by k82cn — APPROVED** [review]
@ravigadde , thanks very much for the new PR; that's OK to me.

**Comment by k82cn** [comment]
merge into one commit.

**Comment by calebamiles** [comment]
Adding `do-not-merge` label until exception is approved by @ethernetdan, @timothysc.

**cc: @kubernetes/kubernetes-release-managers**

**Comment by timothysc** [comment]
@calebamiles exception was already filed, I removed the label too.  Ihor was on the original chain.

**Comment by timothysc** [comment]
@k8s-bot test this #IGNORE

**Comment by timothysc** [comment]
@kubernetes/test-infra-maintainers verify looks to be failing to pull.. looks like an infra issue.

**Comment by timothysc** [comment]
tests aren't passing...

**Comment by ixdy** [comment]
@timothysc which build is failing to pull?

**Comment by timothysc** [comment]
@ixdy verification.

**Comment by ixdy** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41447/pull-kubernetes-verify/20485/ failed on `verify-bazel`:
```
I0301 19:15:07.364] Verifying hack/make-rules/../../hack/verify-bazel.sh
I0301 19:15:13.162] 21a22
I0301 19:15:13.163] >         "//vendor:k8s.io/apimachinery/pkg/types",
I0301 19:15:13.170] 22a23
I0301 19:15:13.170] >         "//vendor:k8s.io/apimachinery/pkg/types",
W0301 19:15:13.271] DRY-RUN: wrote "plugin/pkg/scheduler/api/BUILD"
W0301 19:15:13.271] DRY-RUN: wrote "plugin/pkg/scheduler/api/v1/BUILD"
W0301 19:15:13.350] 
W0301 19:15:13.351] 2 BUILD files not up-to-date.
I0301 19:15:13.451] 
I0301 19:15:13.451] Run ./hack/update-bazel.sh
I0301 19:15:13.451] FAILED   hack/make-rules/../../hack/verify-bazel.sh	6s
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41447/pull-kubernetes-verify/20556/ and https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41447/pull-kubernetes-verify/20560/ both appear to have failed on `go vet`:
```
W0302 07:36:52.562] plugin/pkg/scheduler/core/extender_test.go:175: unreachable code
```

I don't understand what I'm missing.

**Comment by ravigadde** [comment]
@ixdy Thanks, I am looking at it

**Comment by ravigadde** [comment]
@timothysc Can you please retrigger the tests?

**Comment by timothysc** [comment]
@k8s-bot test this #IGNORE

**Comment by davidopp** [comment]
@k8s-bot gci gke e2e test this

**Comment by davidopp** [comment]
@k8s-bot cvm gke e2e test this

**Comment by davidopp** [comment]
@k8s-bot gci gce e2e test this

**Comment by ravigadde** [comment]
@davidopp @timothysc the 2 failures seem unrelated (atleast from what I can see in the logs). Should I try a rebase?

**Comment by k82cn** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this

**Comment by ravigadde** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this

**Comment by ravigadde** [comment]
Rebased.. it will be great if any of you can retrigger tests

@k82cn @davidopp @timothysc

**Comment by k82cn** [comment]
@k8s-bot test this #IGNORE

**Comment by ravigadde** [comment]
I dont understand what the failure is:

I0303 19:09:54.214] Ran 164 of 212 Specs in 1782.522 seconds
I0303 19:09:54.214] SUCCESS! -- 164 Passed | 0 Failed | 0 Flaked | 0 Pending | 48 Skipped 
I0303 19:09:54.214] 
I0303 19:09:54.214] Ginkgo ran 1 suite in 29m45.287562542s
I0303 19:09:54.214] Test Suite Passed
I0303 19:09:54.214] 
I0303 19:09:54.214] Success Finished Test Suite on Host tmp-node-e2e-6f15cd87-gci-stable-56-9000-84-2
I0303 19:09:54.214] <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
I0303 19:09:54.214] <                              FINISH TEST                               <
I0303 19:09:54.215] <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
I0303 19:09:54.215] 
I0303 19:09:54.215] Failure: 1 errors encountered.
W0303 19:09:54.315] exit status 1
W0303 19:09:54.946] Traceback (most recent call last):
W0303 19:09:54.946]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 128, in <module>
W0303 19:09:54.947]     ARGS.service_account,
W0303 19:09:54.947]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 92, in main
W0303 19:09:54.947]     img,
W0303 19:09:54.947]   File "/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/./test-infra/jenkins/../scenarios/kubernetes_kubelet.py", line 44, in check
W0303 19:09:54.947]     subprocess.check_call(cmd)
W0303 19:09:54.947]   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
W0303 19:09:54.947]     raise CalledProcessError(retcode, cmd)
W0303 19:09:54.948] subprocess.CalledProcessError: Command '('docker', 'run', '--rm=true', '-v', '/etc/localtime:/etc/localtime:ro', '-v', '/var/run/docker.sock:/var/run/docker.sock', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/go/src/k8s.io/kubernetes:/go/src/k8s.io/kubernetes', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/go/src/k8s.io/kubernetes/_artifacts:/workspace/_artifacts', '-v', '/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri@tmp/secretFiles/3888997b-7871-4a2a-aaf3-29dba03dc5cb/kubernetes-jenkins-pull-c67f9519f993.json:/service-account.json:ro', '-v', '/home/jenkins/.ssh/google_compute_engine:/root/.ssh/google_compute_engine:ro', '-v', '/home/jenkins/.ssh/google_compute_engine.pub:/root/.ssh/google_compute_engine.pub:ro', '-e', 'GCE_USER=jenkins', '-e', 'GOOGLE_APPLICATION_CREDENTIALS=/service-account.json', '-e', 'JENKINS_GCE_SSH_PRIVATE_KEY_FILE=/root/.ssh/google_compute_engine', '-e', 'JENKINS_GCE_SSH_PUBLIC_KEY_FILE=/root/.ssh/google_compute_engine.pub', '-e', 'NODE_TEST_PROPERTIES=test/e2e_node/jenkins/cri_validation/jenkins-pull.properties', '-e', 'NODE_TEST_SCRIPT=./test/e2e_node/jenkins/e2e-node-jenkins.sh', '-e', 'REPO_DIR=/var/lib/jenkins/workspace/pull-kubernetes-node-e2e-non-cri/go/src/k8s.io/kubernetes', 'gcr.io/k8s-testimages/kubekins-node:1.6-latest')' returned non-zero exit status 1
E0303 19:09:54.953] Build failed

**Comment by davidopp** [comment]
The test failures (I clicked on the three links) seem to be flakes? I'm not sure, we can re-run the tests.

BTW I don't have time to review this carefully right now, but I noticed you decided to stop allowing multiple extenders. The scheduler extension API doesn't say it's alpha or beta, so people might be considering that it is GA (or at the very least it is not alpha). I think this means you need to continue to support the old API, i.e. multiple extenders, for backward compatibility (of course you don't need to support Bind method when people are using the old API).

**Comment by davidopp** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot non-cri node e2e test this

**Comment by k82cn** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this

File: examples/scheduler-policy-config-with-extender.json
Comment (RenaudWasTaken) [inline examples/scheduler-policy-config-with-extender.json]: Is there a design doc that explains the rational for the loss of multiple extenders :) ?

Comment (ravigadde) [inline examples/scheduler-policy-config-with-extender.json]: @RenaudWasTaken not aware of anyone using multiple extenders. Do you have a use case for it? Please see discussion on this PR about getting bind implemented.

**Comment by ravigadde** [comment]
@davidopp Please share your thoughts on this question I asked earlier when you get a chance. To implement Bind the way kube-mesos wants it (delegate Bind to extender as opposed to inform the extender of the Bind), we need to pick option #1 or #2 below.

1) Drop support for multiple extenders and implement Bind the way kube-mesos wants it. Not sure if anyone uses multiple extenders. But your point about not breaking backward compatibility is valid.
2) Check that only one extender can implement Bind method and offload Bind to that extender.
3) Keep Bind the same way it was originally implemented in this PR and invent another mechanism to satisfy the kube-mesos case.

Originally I implemented #3. Based on @timothysc's feedback, switched to #1. #2 may satisfy both concerns. I can implement in such a way that all extenders can be informed of the Bind, but atmost one is delegated to do the Bind with apiserver.

**Comment by k82cn** [comment]
so this will be 1.7?

**Comment by davidopp** [comment]
> so this will be 1.7?

Yes, it's way too late for 1.6 and we don't add new features in patch releases, only bug fixes.

Sorry @ravigadde ...

**Comment by davidopp** [comment]
Sorry I haven't been following this carefully, did you decide to change extender so you can only have one per scheduler? I think this is a breaking change so probably not good. As an aside, @RenaudWasTaken has some use case that I believe involves multiple extenders. which he mentioned at the Resource Management Working Group meeting earlier today.

**Comment by RenaudWasTaken** [comment]
As a quick follow up on what @davidopp is saying, we believe that it might be interesting to use the extenders as "custom resource managers".
The idea being that you would have one per resource (e.g: GPU, FPGA, ...).

The "full" picture is here:
https://groups.google.com/forum/#!topic/kubernetes-sig-scheduling/a-nvvjjP70M

**Comment by RenaudWasTaken** [comment]
Reading this PR and the related issue (https://github.com/kubernetes/kubernetes/issues/41235), it looks like this is exactly what is needed for my use case.

However, @ravigadde solution #3 or #2 seems to fit more in my use case.

Additionally, it looks like the race condition that @ravigadde and @davidopp talked about would still present if #2 was chosen (as the Bind is ran in a goroutine).

Example:
1) Extender's Filter and Prioritize are run
2) Scheduler starts goroutine to bind

3) Scheduler starts to Process next Pod
4) Scheduler calls Extender's Filter
5) Scheduler calls Extender's Bind

**Comment by k82cn** [comment]
@RenaudWasTaken , the race condition is always here :). The point is about error handling, for example, if 3 extender, how to recover states if one of them failed. # 2 provide a central place for developer to keep sync. Cached states introduced several issue for kube-mesos.

**Comment by ravigadde** [comment]
@RenaudWasTaken @davidopp @k82cn  Sorry for the delay in responding. I will implement Option #2 for 1.7, preserving support for multiple extenders.

@RenaudWasTaken If the extender serializes Binds, there is no race condition to deal with. If you have a specific situation in mind, please describe it with an example.

**Comment by ravigadde** [comment]
Working on the 1.7 changes, which I will post in a few days

**Comment by ravigadde** [comment]
Submitted #44883 , closing this now.

**Pull request closed without merging**
